File: src/fr/adrienbrault/idea/symfony2plugin/dic/intention/PhpServiceArgumentIntention.java
package fr.adrienbrault.idea.symfony2plugin.dic.intention;

import com.intellij.codeInsight.hint.HintManager;
import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.psi.xml.XmlTag;
import com.intellij.ui.components.JBList;
import com.intellij.util.IncorrectOperationException;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import fr.adrienbrault.idea.symfony2plugin.action.ServiceActionUtil;
import fr.adrienbrault.idea.symfony2plugin.stubs.ContainerCollectionResolver;
import fr.adrienbrault.idea.symfony2plugin.stubs.ServiceIndexUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.yaml.psi.YAMLKeyValue;

import java.util.*;

public class PhpServiceArgumentIntention extends PsiElementBaseIntentionAction {
    @Override
    public void invoke(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) throws IncorrectOperationException {
        PhpClass phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass.class);
        if(phpClass != null) {
            Set<String> serviceNames = ContainerCollectionResolver.ServiceCollector.create(project).convertClassNameToServices(phpClass.getFQN());
            if(serviceNames.size() > 0) {
                Collection<PsiElement> psiElements = new ArrayList<>();
                for(String serviceName: serviceNames) {
                    psiElements.addAll(ServiceIndexUtil.findServiceDefinitions(project, serviceName));
                }

                if(psiElements.size() > 0) {
                    Map<String, PsiElement> map = new HashMap<>();

                    for (PsiElement element : psiElements) {
                        map.put(VfsUtil.getRelativePath(element.getContainingFile().getVirtualFile(), element.getProject().getBaseDir()), element);
                    }

                    final JBList<String> list = new JBList<>(map.keySet());

                    JBPopupFactory.getInstance().createListPopupBuilder(list)
                        .setTitle("Symfony: Services Definitions")
                        .setItemChoosenCallback(() -> new WriteCommandAction.Simple(editor.getProject(), "Service Update") {
                            @Override
                            protected void run() {
                                String selectedValue = list.getSelectedValue();

                                PsiElement target = map.get(selectedValue);
                                invokeByScope(target, editor);
                            }
                        }.execute())
                        .createPopup()
                        .showInBestPositionFor(editor);
                }
            }
        }
    }

    private void invokeByScope(@NotNull PsiElement psiElement, @NotNull Editor editor) {
        if(psiElement instanceof XmlTag) {
            List<String> args = ServiceActionUtil.getXmlMissingArgumentTypes((XmlTag) psiElement, true, new ContainerCollectionResolver.LazyServiceCollector(psiElement.getProject()));
            if (args == null) {
                HintManager.getInstance().showErrorHint(editor, "No argument update needed");
                return;
            }

            ServiceActionUtil.fixServiceArgument(args, (XmlTag) psiElement);
        } else if(psiElement instanceof YAMLKeyValue) {
            ServiceActionUtil.fixServiceArgument((YAMLKeyValue) psiElement);
        }
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) {
        return true;
    }

    @Nls
    @NotNull
    @Override
    public String getFamilyName() {
        return "Symfony";
    }

    @NotNull
    @Override
    public String getText() {
        return "Symfony: Update service arguments";
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/dic/intention/PhpServiceArgumentIntention.kt
package fr.adrienbrault.idea.symfony2plugin.dic.intention

import com.intellij.codeInsight.hint.HintManager
import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
import com.intellij.codeInsight.intention.preview.IntentionPreviewInfo
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.xml.XmlTag
import com.jetbrains.php.lang.PhpLanguage
import com.jetbrains.php.lang.psi.elements.PhpClass
import fr.adrienbrault.idea.symfony2plugin.action.ServiceActionUtil
import fr.adrienbrault.idea.symfony2plugin.stubs.ContainerCollectionResolver
import fr.adrienbrault.idea.symfony2plugin.stubs.ServiceIndexUtil
import fr.adrienbrault.idea.symfony2plugin.util.IdeHelper
import fr.adrienbrault.idea.symfony2plugin.util.ProjectUtil
import org.jetbrains.yaml.psi.YAMLKeyValue

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class PhpServiceArgumentIntention : PsiElementBaseIntentionAction() {
    override fun invoke(project: Project, editor: Editor?, psiElement: PsiElement) {
        val phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass::class.java)
            ?: return

        val serviceNames = ContainerCollectionResolver.ServiceCollector.create(project)
            .convertClassNameToServices(phpClass.fqn)
        if (serviceNames.isEmpty()) {
            return
        }

        val psiElements = ArrayList<PsiElement>()
        for (serviceName in serviceNames) {
            psiElements.addAll(ServiceIndexUtil.findServiceDefinitions(project, serviceName))
        }

        if (psiElements.isEmpty()) {
            return
        }

        val map = HashMap<String?, PsiElement>()

        for (element in psiElements) {
            map[VfsUtil.getRelativePath(element.containingFile.virtualFile, ProjectUtil.getProjectDir(element))] = element
        }

        JBPopupFactory.getInstance().createPopupChooserBuilder(ArrayList(map.keys))
            .setTitle("Symfony: Services Definitions")
            .setItemChosenCallback { selectedValue ->
                WriteCommandAction.writeCommandAction(project)
                    .withName("Service Update")
                    .run<RuntimeException> { invokeByScope(project, map[selectedValue]!!, editor!!) }
            }
            .createPopup()
            .showInBestPositionFor(editor!!)
    }

    private fun invokeByScope(project: Project, psiElement: PsiElement, editor: Editor) {
        var success = false

        if (psiElement is XmlTag) {
            val args = ServiceActionUtil.getXmlMissingArgumentTypes(
                psiElement,
                true,
                ContainerCollectionResolver.LazyServiceCollector(psiElement.project),
            )

            success = args.isNotEmpty()
            if (success) {
                ServiceActionUtil.fixServiceArgument(args, psiElement)
            }
        } else if (psiElement is YAMLKeyValue) {
            val args = ServiceActionUtil.getYamlMissingArgumentTypes(
                project,
                ServiceActionUtil.ServiceYamlContainer.create(psiElement),
                false,
                ContainerCollectionResolver.LazyServiceCollector(psiElement.project),
            )

            success = args.isNotEmpty()
            if (success) {
                ServiceActionUtil.fixServiceArgument(psiElement)
            }
        }

        if (!success) {
            IdeHelper.showErrorHintIfAvailable(editor, "No argument update needed")
            return
        }

        var relativePath = VfsUtil.getRelativePath(
            psiElement.containingFile.virtualFile,
            ProjectUtil.getProjectDir(psiElement),
        )
        if (relativePath == null) {
            relativePath = "n/a"
        }

        HintManager.getInstance().showInformationHint(editor, String.format("Argument updated: %s", relativePath))
    }

    override fun isAvailable(project: Project, editor: Editor?, psiElement: PsiElement): Boolean {
        return psiElement.language == PhpLanguage.INSTANCE && getServicesInScope(project, psiElement).isNotEmpty()
    }

    override fun getFamilyName(): String {
        return "Symfony: Update service arguments"
    }

    override fun getText(): String {
        return familyName
    }

    private fun getServicesInScope(project: Project, psiElement: PsiElement): Set<String> {
        val phpClass = PsiTreeUtil.getParentOfType(psiElement, PhpClass::class.java)

        return if (phpClass == null) {
            emptySet()
        } else {
            ContainerCollectionResolver.ServiceCollector.create(project).convertClassNameToServices(phpClass.fqn)
        }
    }

    override fun generatePreview(project: Project, editor: Editor, file: PsiFile): IntentionPreviewInfo {
        return IntentionPreviewInfo.EMPTY
    }
}
